tree2: Add schema 2 proposal (#16217)
microsoft/FluidFramework: experimental/dds/tree2/docs/schema2.md

5. It requires use of symbols to avoid colliding with local field names in many places in the API, which has been confusing users (many TypeScript developers don't even know what Symbols are).
6. Supporting all the things in all the nodes makes APIs complex. This is extra annoying since this complicates the schema-aware API.
7. Extra fields require proxies to implement the desired TypeScript APIs for nodes.
8. Having both global and local fields adds complexity, particularly to schema and path storage where symbols can't be used.
9. Supporting extraLocalFields and extraGlobalFields on nodes while providing a JavaScript object like API requires a Proxy, which adds overhead and complicates code implementation and maintenance.

## Proposal

Instead of one kind of node in view schema, have 4, each with a subset of our current functionality:

1. Terminal Node: holds a value, but no children.
   Don't support `undefined` as a value: an empty Struct can be used for that case if needed.
2. Struct Node: finite list of fields (key+field type).

> 3. Map Node: node with a single single field type which must permit empty.

Reviewer: I assume the "single single" in an unintended repetition.

I don't really understand what "single field type which must permit empty" means here. Are you saying that Map nodes would be parameterized with a field kind whose multiplicity would have to tolerate zero children? For example, one could make a map of optionals, or a map of sequences. If that is what you mean then maybe "field kind" would be clearer since "field type" could be referring to the type of children allowed in a field.

Author: should be more clear/fixed now.

> 3. Map Node: node with a single single field type which must permit empty.

Reviewer: Note that we don't currently have a "clear" operation that could be performed on this map node. Something we could look into adding though.

   Allows any field of that type under any string field key.
   Provides the functionality currently done by making a node schema with extra local fields.

> 4. Field Node: Has a single unnamed filed (using the empty field key).

Reviewer: There are two concepts here:
1. A node that's parameterized with a field kind and implicitly unwraps it. I think "Transparent Node" may be a better name for this.
2. A specific instantiation of 1. Right now, it looks like sequence field is the only field that would benefit from this, so we'd likely call that instantiation "SequenceNode".

Author: There are use cases for doing this with optional fields: maybe not super high value usecases. For example, you could have a field which takes a color, and implement a color which can be transparent as an optional color node. This could be useful in a context where you want to take in that optional value or something else, like a button that that an optional color (which can be transparent) or an image.

I'm not sure Transparent Node is a great name since terminal nodes are transparent (and just show the value) by default, where field nodes do the same for a field.

Reviewer: I'm not sure I fully follow the example. Describing it using TS types, are you saying that it lets us have

`OptionalColor = { [""]: Color | undefined }`
`ButtonAppearance = OptionalColor | Image`

instead of having

`OptionalColor = Color | undefined`
`ButtonAppearance = OptionalColor | Image` which is `Color | undefined | Image`

It's not clear to me in that example what value we get from the unwrapping.

Your point about the name "transparent" makes sense to me, and I don't have a better suggestion, so Field Node it is. I do think having `SequenceNode` be a thing and making sure most users interact with that concept without being forced to learn about `FieldNode` would be good, but I understand that's not this document's primary concern.

Author: Yes, thats the example. In that example, its different from having a single optional Image or Color since the optional color can have a merge identity even when its transparent, and can be edited to be not transparent. If you app has a color picker projected from the optional color for example that might work better. Also if image ws optional, and you wanted to be able to tell no image apart from transparent color, you would need some optional node pattern like this.

Reviewer: I think that motivates having a wrapping layer. That part makes sense to me, it's the implicit unwrapping that doesn't make sense to me in this scenario.

   When reading in the editable tree API implicitly, unwraps to the field.
   This provides the functionality currently done by making a node schema with a primary field (empty field key), which used for "array nodes".

Each of these will get a separate API to declare (similar to SchemaBuilder.object and SchemaBuilder.primitive, except we will have 4 options).
Each will also get its own API for accessing it in editable tree, though all will extend a common generic tree traversal API as well.

This will mirror how field kinds work, where we have a small collection of supported kinds in schema, and each gets its own editable tree API.

Struct Nodes will also get an additional feature: custom field names.
Custom field names allow the view schema to declare the string to be used for the name of the field in schema-aware APIs instead of just using the field key.
This can default to the field key, but can optionally be a distinct string.

> Customizable field names are great for app maintainability: a field in coded can be renamed as used in code without breaking existing data.

Reviewer: There's something wonky with the "a field in coded can be renamed as used in code" part

There are also some further benefits:
it makes it practical to ban specific field names for use in APIs if they collide with our framework APIs.
Since an app can rename the field without changing how data is persisted, requiring them to do this remaining if a framework update causes a name collision will be ok.
This means Struct Nodes can have scheme-aware APIs that don't have to use symbols to avoid field name collisions.
We will keep a method to look up fields by their keys to cover generic code as well.

Custom field names make it practical to to use long collision resistant names for fields, and this usage pattern can replace most of the need for global field keys.
The other use case, where the same field is desired on multiple schema can be handled in other ways, like putting it on a reused child node, or just putting the field directly on each schema its desired on.

Map nodes will not expose a JavaScript object like API, and instead expose a JavaScript Map like API.
This avoids the need to use a Proxy for any of the node implementations, as well as avoids needing to support custom field names for Map nodes to deal with possible API name collisions.

### Annotation Pattern

> ExtraGlobalFields in the previous design existed to support an "annotations pattern" where a an app could opt some or all of their schema into allowing annotation subtrees to be placed on nodes.

Reviewer: ```suggestion
ExtraGlobalFields in the previous design existed to support an "annotations pattern" where an app could opt some or all of their schema into allowing annotation subtrees to be placed on nodes.
```

ExtraGlobalFields isn't really a full solution to this.
The real challenge is that different applications may have different view schema, each with a different set of known annotations.
Our stored vs view schema already can model this, even without ExtraGlobalFields:
each application can add their annotations as optional fields to the stored schema.

The only thing thats really missing is how the applications should handle opening documents with stored fields like these.
This an be addressed by supporting either one or both of the following:

1.  Allow field in the schema (stored and view) to indicate how applications which do not understand the field should treat the type containing the field.
    Some options are:
    1.  read+write, preserve unknown field where possible
    1.  read+write, clear unknown field on mutation
    1.  readonly
    1.  Treat as out of schema / unsupported
2.  Allow nodes to declare if they support unrecognized fields (in the view schema only).
    This would generate an API that could enumerate unexpected fields in the schema-aware API.

Additionally an API could be added to struct nodes to enumerate all unexpected fields.

> API wise this looks similar to the existing extra fields support, but its distinct in its use-case and performance characteristics.

Reviewer: ```suggestion
API wise this looks similar to the existing extra fields support, but it is distinct in its use-case and performance characteristics.
```

Extra local fields were designed to allow allow arbitrary fields, without bloating the stored schema:

> in the previous system adding and removing N extra local fields with different keys currently makes no schema changes where as adding a bunch of fields to the stored scheme as unrecognized fields in the proposed schema system would bloat the document schema proportionally.

Reviewer: ```suggestion
in the previous system adding and removing N extra local fields with different keys currently makes no schema changes whereas adding a bunch of fields to the stored schema as unrecognized fields in the proposed schema system would bloat the document schema proportionally.
```
